File: android/app/src/main/java/se/pixelmug/glucoseface/Alerts.kt
package se.pixelmug.glucoseface

import kotlin.math.max
import kotlin.math.min

/** Level + expression from a glucose series, mirroring alerts.ts / face.ts. */
object Alerts {
    const val LOW_URGENT = 3.0
    const val HIGH_URGENT = 14.0
    const val PRED_WINDOW_MIN = 20.0

    data class Assessment(val level: String, val expr: String, val current: Double, val predicted: Double)

    /** Least-squares slope (mmol/min) over the last 20 min, dedup + span guarded. */
    private fun slopePerMin(readings: List<Reading>): Double {
        if (readings.size < 2) return 0.0
        val tEnd = readings.last().ts
        val seen = HashSet<Long>()
        val pts = readings.filter { tEnd - it.ts <= 20 * 60 * 1000 }.filter { seen.add(it.ts) }
        if (pts.size < 2) return 0.0
        val t0 = pts.first().ts
        if ((tEnd - t0) / 60000.0 < 5) return 0.0
        val xs = pts.map { (it.ts - t0) / 60000.0 }
        val ys = pts.map { it.mmol }
        val n = pts.size
        val sx = xs.sum(); val sy = ys.sum()
        val sxx = xs.sumOf { it * it }
        val sxy = xs.indices.sumOf { xs[it] * ys[it] }
        val denom = n * sxx - sx * sx
        if (denom <= 1e-6) return 0.0
        return max(-1.0, min(1.0, (n * sxy - sx * sy) / denom))
    }

    fun assess(readings: List<Reading>, nowMs: Long): Assessment {
        if (readings.isEmpty()) return Assessment("unknown", "sleep", Double.NaN, Double.NaN)
        val last = readings.last()
        val cur = last.mmol
        val ageMin = (nowMs - last.ts) / 60000.0
        if (cur <= 0) return Assessment("unknown", "sleep", cur, Double.NaN)
        val slope = slopePerMin(readings)
        val pred = max(1.0, min(30.0, cur + slope * PRED_WINDOW_MIN))
        if (ageMin > 16) return Assessment("stale", "sleep", cur, pred)

        val level = when {
            cur <= LOW_URGENT -> "urgentLow"
            cur >= HIGH_URGENT -> "urgentHigh"
            slope < 0 && pred <= LOW_URGENT -> "predLow"
            slope > 0 && pred >= HIGH_URGENT -> "predHigh"
            else -> "ok"
        }
        return Assessment(level, expressionForLevel(level), cur, pred)
    }

    fun expressionForLevel(level: String): String = when (level) {
        "urgentLow" -> "panic"
        "predLow" -> "worried"
        "predHigh" -> "queasy"
        "urgentHigh" -> "sick"
        "stale", "unknown" -> "sleep"
        else -> "happy"
    }
}

File: src/alerts.ts
/**
 * alerts.ts — glucose warnings + 20-minute prediction for the PixelMug.
 *
 * Two kinds of warning, both surfaced as scrolling coloured text on the mug:
 *   - ACUTE   : you are already too low/high  ("UNDER 3 – ÄT NU!", "ÖVER 14!")
 *   - PREDICT : the 20-min projection will cross a limit ("SNART UNDER 3 …")
 *
 * Prediction = current + slope × window, where slope is the least-squares trend
 * over the last 20 min (not Dexcom's arrow, which lags real falls).
 */
import { slopePerMin, type Reading } from "./dexcom";

export const RED = "#ff2b2b";
export const AMBER = "#ff9500";

export type AlertThresholds = {
  lowUrgent: number; // at/below -> acute low
  highUrgent: number; // at/above -> acute high
  predWindowMin: number; // how far ahead to project
};

export const DEFAULT_THRESHOLDS: AlertThresholds = {
  lowUrgent: 3.0,
  highUrgent: 14.0,
  predWindowMin: 20,
};

export type Level =
  | "unknown"
  | "stale"
  | "urgentLow"
  | "predLow"
  | "ok"
  | "predHigh"
  | "urgentHigh";

export type Assessment = {
  current: number;
  ageMin: number;
  slope: number; // mmol/min
  predicted: number; // mmol projected `predWindowMin` ahead
  level: Level;
  warning: { text: string; color: string } | null;
};

export function assess(
  readings: Reading[],
  t: AlertThresholds = DEFAULT_THRESHOLDS,
  nowTs?: number,
): Assessment {
  const empty = (level: Level): Assessment => ({
    current: NaN,
    ageMin: NaN,
    slope: 0,
    predicted: NaN,
    level,
    warning: null,
  });

  if (!readings.length) return empty("unknown");
  const last = readings[readings.length - 1];
  const now = nowTs ?? last.ts;
  const current = last.mmol;
  const ageMin = (now - last.ts) / 60000;
  if (Number.isNaN(current) || current <= 0) return empty("unknown");

  const slope = slopePerMin(readings, 20);
  const predicted = Math.max(1, Math.min(30, current + slope * t.predWindowMin));

  // Stale data can't be trusted for a live warning.
  if (ageMin > 16) {
    return { current, ageMin, slope, predicted, level: "stale", warning: null };
  }

  let level: Level = "ok";
  let warning: Assessment["warning"] = null;
  const win = t.predWindowMin;

  if (current <= t.lowUrgent) {
    level = "urgentLow";
    warning = { text: `LÅGT ${current.toFixed(1)} – ÄT NU`, color: RED };
  } else if (current >= t.highUrgent) {
    level = "urgentHigh";
    warning = { text: `HÖGT ${current.toFixed(1)}!`, color: RED };
  } else if (slope < 0 && predicted <= t.lowUrgent) {
    level = "predLow";
    warning = { text: `SNART LÅGT – ~${predicted.toFixed(1)} om ${win} min`, color: RED };
  } else if (slope > 0 && predicted >= t.highUrgent) {
    level = "predHigh";
    warning = { text: `SNART HÖGT – ~${predicted.toFixed(1)} om ${win} min`, color: AMBER };
  }

  return { current, ageMin, slope, predicted, level, warning };
}
